workflow/scripts/osemosys_global/powerplant/user_defined_capacity.py

```diff
@@ -0,0 +1,255 @@
+                                if not df_res_cap_ud_final.empty else None])
+                
+        # For technologies with start year at or after model start year, add to 
+        # TotalAnnualMinCapacityInvestment      
+        df_min_cap_inv = df_min_cap_inv.loc[df_min_cap_inv['YEAR'] >= min(get_years(start_year, end_year))]
+        df_min_cap_inv.drop_duplicates(subset=['REGION', 
+                                               'TECHNOLOGY',
+                                               'YEAR'],
+                                       keep='last',
+                                       inplace=True)
+        
+        tech_list = list(tech_capacity_df['TECHNOLOGY'].unique())
+       
+        df_iar_custom = pd.DataFrame(list(itertools.product(tech_list,
+                                                        [1, 2],
+                                                        get_years(start_year, end_year))
+                                      ),
+                                 columns = ['TECHNOLOGY',
+                                            'MODE_OF_OPERATION',
+                                            'YEAR']
+                                 )
+        df_oar_custom = pd.DataFrame(list(itertools.product(tech_list,
+                                                        [1, 2],
+                                                        get_years(start_year, end_year))
+                                      ),
+                                 columns = ['TECHNOLOGY',
+                                            'MODE_OF_OPERATION',
+                                            'YEAR']
+                                 )  
+        # IAR in modes 1 and 2 are primary electricity commodity ('ELC*01') in 
+        # node_from and node_to, respectively. 
+        # OAR is the inverse of the above
+        df_iar_custom.loc[df_iar_custom['MODE_OF_OPERATION']==1,
+                                        'FUEL'] = ('ELC' + 
+                                                   df_iar_custom['TECHNOLOGY'].str[3:8] + 
+                                                   '01')
+        df_iar_custom.loc[df_iar_custom['MODE_OF_OPERATION']==2,
+                                        'FUEL'] = ('ELC' + 
+                                                   df_iar_custom['TECHNOLOGY'].str[8:13] + 
+                                                   '01')
+        df_oar_custom.loc[df_iar_custom['MODE_OF_OPERATION']==1,
+                                        'FUEL'] = ('ELC' + 
+                                                   df_oar_custom['TECHNOLOGY'].str[8:13] + 
+                                                   '01')
+        df_oar_custom.loc[df_iar_custom['MODE_OF_OPERATION']==2,
+                                        'FUEL'] = ('ELC' + 
+                                                   df_oar_custom['TECHNOLOGY'].str[3:8] + 
+                                                   '01')
```

Comment:
Im confused about whats happening here! Are these fuel types being assigned correctly? 

Input activity ratios should be primary fuel codes? 

![image](https://github.com/user-attachments/assets/90a31444-599f-4c6a-9ece-c151ae88ea7b)

Output activity ratios should be end-use fuel codes? 

![image](https://github.com/user-attachments/assets/d51e1f46-ea2c-4281-b3c2-edf71fa5c243)

Reply:
I believe this is now addressed in [this commit](https://github.com/OSeMOSYS/osemosys_global/pull/206/commits/a85e95bd32bbfbed1d62b832815ad99d1ec4e01d)